### PR TITLE
Signals

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
cc @gkoz

### CODE REVIEW

**Comment by gkoz** [comment]
Can you double-check that leaf classes don't have unnecessary generics and downcasts but `CellRenderer` has a proper generic implementation of the Signals trait?

**Comment by GuillaumeGomez** [comment]
"leaf classes"?

**Comment by EPashkin** [comment]
Trampolines for `CellRendererAccelSignals::connect_accel_cleared<F>` don't need `where T: IsA<CellRendererAccel>` as no subclass for it.
Same for `CellRendererCombo`.
Only `CellRenderer`s trampolines need <T>.
May be you also forgot adding signal `https://developer.gnome.org/gtk3/stable/GtkCellRendererText.html#GtkCellRendererText-edited` (it also need <T> as it base class for `GtkCellRendererAccel`, `GtkCellRendererCombo`, `GtkCellRendererSpin`)

**Comment by EPashkin** [comment]
And `https://developer.gnome.org/gtk3/stable/GtkCellRendererToggle.html#GtkCellRendererToggle-toggled`

**Comment by EPashkin** [comment]
@gkoz also pointed that `CellRenderer` impl need `impl<T: IsA<CellRenderer> + IsA<Object>> super::CellRendererSignals for T {`

**Comment by GuillaumeGomez** [comment]
> And `https://developer.gnome.org/gtk3/stable/GtkCellRendererToggle.html#GtkCellRendererToggle-toggled`

Shouldn't it be put into his own trait?

**Comment by GuillaumeGomez** [comment]
> May be you also forgot adding signal `https://developer.gnome.org/gtk3/stable/GtkCellRendererText.html#GtkCellRendererText-edited` (it also need as it base class for GtkCellRendererAccel, GtkCellRendererCombo, GtkCellRendererSpin)

I didn't, I just walk through structures and implement their signals if we didn't so it will come later.

**Comment by GuillaumeGomez** [comment]
Anyway, updated.

File: src/signal.rs [hunk 1]
```diff
@@ -2093,3 +2094,215 @@ mod cell_area {
...
+mod cell_editable {
+    use CellEditable;
+    use std::mem::transmute;
+    use ffi::GtkCellEditable;
+    use glib::object::Downcast;
+    use glib::signal::connect;
+    use glib::translate::*;
+    use IsA;
+
+    impl super::CellEditableSignals for CellEditable {
```
Comment (EPashkin) [inline src/signal.rs]: Its interface, so it need be `T:IsA<CellEditable>+IsA<Object>`

**Comment by EPashkin** [comment]
`GtkCellRendererToggle` need trait `CellRendererToggleSignal` but as has no children don't need <T> in impl and trampolines.
Adding `GtkCellRendererText` signal will cover last part of `CellRenderer` tree.
All of it get `TreePath` as string, so we can't currently automatic generate it.

File: src/signal.rs [hunk 2]
```diff
@@ -2093,3 +2094,215 @@ mod cell_area {
...
+            }
+        }
+    }
+
+    unsafe extern "C" fn trampoline(this: *mut GtkCellRendererAccel,
+                                    path: *const c_char,
+                                    f: &Box<Fn(&CellRendererAccel, TreePath) + 'static>) {
+        callback_guard!();
+        let path = from_glib_full(gtk_tree_path_new_from_string(path));
+        f(&CellRendererAccel::from_glib_none(this).downcast_unchecked(), path);
```
Comment (EPashkin) [inline src/signal.rs]: `.downcast_unchecked()` unneeded when no generalisation.

**Comment by GuillaumeGomez** [comment]
It'll be part of another PR I guess. :)

File: src/signal.rs [hunk 3]
```diff
@@ -2093,3 +2094,216 @@ mod cell_area {
...
+            }
+        }
+    }
+
+    unsafe extern "C" fn trampoline(this: *mut GtkCellRendererCombo,
+                                    path: *const c_char,
+                                    f: &Box<Fn(&CellRendererCombo, TreePath) + 'static>) {
+        callback_guard!();
+        let path = from_glib_full(gtk_tree_path_new_from_string(path));
+        f(&CellRendererCombo::from_glib_none(this).downcast_unchecked(), path);
```
Comment (EPashkin) [inline src/signal.rs]: Unneeded downcast

File: src/signal.rs [hunk 4]
```diff
@@ -2093,3 +2094,216 @@ mod cell_area {
...
+                connect(self.to_glib_none().0, "accel-cleared",
+                    transmute(trampoline as usize), Box::into_raw(f) as *mut _)
+            }
+        }
+
+        fn connect_accel_edited<F>(&self, accel_edited_func: F) -> u64
+        where F: Fn(&Self, TreePath, u32, ModifierType, u32) + 'static {
+            unsafe {
+                let f: Box<Box<Fn(&Self, TreePath, u32, ModifierType, u32) + 'static>> =
+                Box::new(Box::new(accel_edited_func));
```
Comment (EPashkin) [inline src/signal.rs]: Extra spaces?

**Comment by EPashkin** [comment]
Thanks

**Comment by GuillaumeGomez** [comment]
Fixed and fixed (and thanks to you for the review!).

File: src/signal.rs [hunk 5]
```diff
@@ -2093,3 +2094,216 @@ mod cell_area {
...
+                    Box::new(Box::new(accel_cleared_func));
+                connect(self.to_glib_none().0, "changed",
+                    transmute(trampoline as usize), Box::into_raw(f) as *mut _)
+            }
+        }
+    }
+
+    unsafe extern "C" fn trampoline(this: *mut GtkCellRendererCombo,
+                                    path: *const c_char,
+                                    f: &Box<Fn(&CellRendererCombo, TreePath) + 'static>) {
```
Comment (gkoz) [inline src/signal.rs]: The `GtkTreeIter*` argument seems to be missing: https://developer.gnome.org/gtk3/stable/GtkCellRendererCombo.html#GtkCellRendererCombo-changed

File: src/signal.rs [hunk 6]
```diff
@@ -2093,3 +2094,216 @@ mod cell_area {
...
+                    Box::new(Box::new(editing_started_func));
+                connect(self.to_glib_none().0, "editing-started",
+                    transmute(start_trampoline::<Self> as usize), Box::into_raw(f) as *mut _)
+            }
+        }
+    }
+
+    unsafe extern "C" fn trampoline<T>(this: *mut GtkCellRenderer,
+                                       path: *const c_char,
+                                       f: &Box<Fn(&T, TreePath) + 'static>)
```
Comment (gkoz) [inline src/signal.rs]: I don't see a `path` argument here: https://developer.gnome.org/gtk3/stable/GtkCellRenderer.html#GtkCellRenderer-editing-canceled

**Comment by GuillaumeGomez** [comment]
@gkoz: Corrected.

**Comment by gkoz** [comment]
Thanks!

**Pull request merged**